lib/at_cmd/at_cmd.c

```diff
@@ -171,124 +290,130 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+			memcpy(ret.buf, buf, payload_len);
 		}
 next:
-		/* If no callback was set, free the item.
-		 * Otherwise, work queue callback will free it.
-		 */
-		if (item->callback == NULL) {
-			k_mem_slab_free(&rsp_work_items, (void **)&item);
-		} else {
-			k_work_init(&item->work, callback_worker);
-			k_work_submit(&item->work);
-		}
-
-		/* Notify back only if command was sent. */
-		if ((k_sem_count_get(&cmd_pending) == 0) &&
-		    (ret.state != AT_CMD_NOTIFICATION)) {
-			current_cmd_handler = NULL;
+		distribute_resp(&current_cmd, &ret);
 
-			k_msgq_put(&return_code_msq, &ret, K_FOREVER);
+		/* We have now handled a command if it was not a notification */
+		if (ret.state != AT_CMD_NOTIFICATION) {
+			complete_cmd();
 		}
 	}
 }
 
-static inline int at_write(const char *const cmd, enum at_cmd_state *state)
+/* Allocate and copy cmd to a new buffer */
+static char *buffer_cmd(const char *cmd)
 {
-	int bytes_sent;
-	int bytes_to_send = strlen(cmd);
-	struct return_state_object ret;
-
-	LOG_DBG("Sending command %s", log_strdup(cmd));
-
-	bytes_sent = send(common_socket_fd, cmd, bytes_to_send, 0);
-
-	if (bytes_sent == -1) {
-		LOG_ERR("Failed to send AT command (err:%d)", errno);
-		ret.code  = -errno;
-		ret.state = AT_CMD_ERROR;
-	} else {
-		LOG_DBG("Awaiting response for %s", log_strdup(cmd));
-		k_msgq_get(&return_code_msq, &ret, K_FOREVER);
-		LOG_DBG("Bytes sent: %d", bytes_sent);
-
-		if (bytes_sent != bytes_to_send) {
-			LOG_ERR("Bytes sent (%d) was not the "
-				"same as expected (%d)",
-				bytes_sent, bytes_to_send);
-		}
-	}
+	char *temp;
 
-	if (state) {
-		*state = ret.state;
+	temp = k_malloc(strlen(cmd) + 1);
+	if (temp == NULL) {
+		return NULL;
 	}
-
-	return ret.code;
+	memcpy(temp, cmd, strlen(cmd) + 1);
+	return temp;
 }
 
 int at_cmd_write_with_callback(const char *const cmd,
-			       at_cmd_handler_t  handler,
-			       enum at_cmd_state *state)
+			       at_cmd_handler_t  handler)
 {
-	k_sem_take(&cmd_pending, K_FOREVER);
+	struct cmd_item command;
+	int ret;
 
-	current_cmd_handler = handler;
+	if (cmd == NULL) {
+		return -EINVAL;
+	}
+
+	command.cmd = buffer_cmd(cmd);
+
+	if (command.cmd == NULL) {
+		return -ENOMEM;
+	}
 
-	int return_code = at_write(cmd, state);
+	command.buf = NULL;
+	command.callback = handler;
+	command.flags = AT_CMD_BUF_CMD | AT_CMD_BUF_RESP;
 
-	k_sem_give(&cmd_pending);
+	ret = k_msgq_put(&commands, &command, K_FOREVER);
+	if (ret) {
+		return ret;
+	}
 
-	return return_code;
+	load_cmd_and_write();
+	return 0;
 }
 
 int at_cmd_write(const char *const cmd,
 		 char *buf,
 		 size_t buf_len,
 		 enum at_cmd_state *state)
 {
-	k_sem_take(&cmd_pending, K_FOREVER);
+	struct cmd_item command;
+	struct resp_item ret;
 
-	response_buf     = buf;
-	response_buf_len = buf_len;
+	if (cmd == NULL) {
+		LOG_ERR("cmd is NULL");
+		*state = AT_CMD_ERROR_QUEUE;
+		return -EINVAL;
+	}
+
+	/* This cast is safe; we do not free cmd without AT_CMD_BUF_CMD */
+	command.cmd = (char *)cmd;
+
```

Comment:
Unneeded empty line.

Reply:
I almost always separate a block with empty lines when using a comment, to make it obvious what the comment refers to. I agree it is pretty obvious here, though. Removed.